pandas/core/groupby/grouper.py

```diff
@@ -338,7 +339,11 @@ def _set_grouper(self, obj: FrameOrSeries, sort: bool = False):
             if getattr(self.grouper, "name", None) == key and isinstance(
                 obj, ABCSeries
             ):
-                ax = self._grouper.take(obj.index)
+                indices = group_indices.get(obj.name)
+                if self._indexer is not None:
+                    ax = self._grouper.take(self._indexer.argsort()).take(indices)
+                else:
+                    ax = self._grouper.take(indices)
```

Comment:
If the df was not resorted, indexer is None. We have to use indices, because a for example a string index would kill obj.index